### PR TITLE
librbd: enable/disable of features is not allowed when already enabled/disabled

### PR DESCRIPTION
**Pull request by s09816**
…lready

enabled/disabled

Fixes:#16079

Signed-off-by: Lu Shi <shi.lu@h3c.com>

### CODE REVIEW

File: src/librbd/internal.cc [hunk 1]
```diff
@@ -1701,16 +1701,20 @@ int mirror_image_disable_internal(ImageCtx *ictx, bool force,
       uint64_t new_features;
       if (enabled) {
         features &= ~ictx->features;
+	if (features == 0) {
```
Comment (dillaman) [inline src/librbd/internal.cc]: I think you would need to return -EINVAL if any of the requested feature bits are already enabled.  For example, if object-map is already enabled and the user attempted to enable object-map and fast-diff.

File: src/librbd/internal.cc [hunk 2]
```diff
@@ -1701,16 +1701,20 @@ int mirror_image_disable_internal(ImageCtx *ictx, bool force,
...
+	  lderr(cct) << "the features are already enabled" << dendl;
```
Comment (dillaman) [inline src/librbd/internal.cc]: Minor: perhaps something like "one or more requested features are already enabled"

**Comment by s09816** [comment]
@dillaman I have modified the code according your advise

**Comment by dillaman** [comment]
@s09816 ```unittest_librbd``` seg faulted during the build check

**Comment by dillaman** [comment]
@s09816 Minor: please update your commit message to use "Fixes: <URL to tracker ticket" format

File: src/librbd/internal.cc [hunk 3]
```diff
@@ -1700,17 +1700,19 @@ int mirror_image_disable_internal(ImageCtx *ictx, bool force,
...
-        features &= ~ictx->features;
+	if (features & ictx->features != 0) {
+	  lderr(cct) << "one or more requested features are already enabled" << dendl;
+	  return -EINVAL;
+        }
         new_features = ictx->features | features;
       } else {
-        features &= ictx->features;
+	if (features & ~ictx->features != 0) {
+	  lderr(cct) << "one or more requested features are not enabled" << dendl;
```
Comment (dillaman) [inline src/librbd/internal.cc]: "one or more requested features are already disabled"

File: src/librbd/internal.cc [hunk 4]
```diff
@@ -1700,17 +1700,19 @@ int mirror_image_disable_internal(ImageCtx *ictx, bool force,
       RWLock::WLocker snap_locker(ictx->snap_lock);
       uint64_t new_features;
       if (enabled) {
-        features &= ~ictx->features;
```
Comment (dillaman) [inline src/librbd/internal.cc]: Still need these masking ops ... just move them to after your new check

**Comment by s09816** [comment]
@dillaman  I have updated.

File: src/librbd/internal.cc [hunk 5]
```diff
@@ -1700,17 +1700,21 @@ int mirror_image_disable_internal(ImageCtx *ictx, bool force,
       RWLock::WLocker snap_locker(ictx->snap_lock);
       uint64_t new_features;
       if (enabled) {
+      	if (features & ictx->features != 0) {
```
Comment (dillaman) [inline src/librbd/internal.cc]: You need to put braces around ```features & ictx->features``` -- similar comment below.  Right now the rbd CLI isn't throwing the error.

**Comment by s09816** [comment]
@dillaman I have updated

**Comment by zhouyuan** [comment]
@s09816 There might be a few tests on update_features need to be updated also, since the return val is `-EINVAL` in some cases. Like this:
https://github.com/ceph/ceph/blob/master/src/test/librbd/test_mirroring.cc#L174

**Comment by dillaman** [comment]
lgtm

**Comment by s09816** [comment]
@dillaman why checks failed

**Comment by dillaman** [comment]
Thanks -- missed that since I thought it was the usual 'unittest_log' segfault. Let me check ...

**Comment by dillaman** [comment]
Run ```test/run-rbd-unit-tests.sh``` to repeat the failure:

> RBD_FEATURES=61 ./unittest_librbd --gtest_filter=TestMockImageRefreshRequest.*
> ...
> [ RUN      ] TestMockImageRefreshRequest.DisableExclusiveLock
> test/librbd/image/test_mock_RefreshRequest.cc:510: Failure
> Value of: update_features(ictx, (1<<2) | (1<<3) | (1<<4) | (1<<6), false)
>   Actual: -22
> Expected: 0
> [  FAILED  ] TestMockImageRefreshRequest.DisableExclusiveLock (3 ms)
> ...
> [ RUN      ] TestMockImageRefreshRequest.EnableObjectMapWithExclusiveLock
> test/librbd/image/test_mock_RefreshRequest.cc:648: Failure
> Value of: update_features(ictx, (1<<6), false)
>   Actual: -22
> Expected: 0
> [  FAILED  ] TestMockImageRefreshRequest.EnableObjectMapWithExclusiveLock (4 ms)
> [ RUN      ] TestMockImageRefreshRequest.EnableObjectMapWithoutExclusiveLock
> test/librbd/image/test_mock_RefreshRequest.cc:682: Failure
> Value of: update_features(ictx, (1<<6), false)
>   Actual: -22
> Expected: 0
> [  FAILED  ] TestMockImageRefreshRequest.EnableObjectMapWithoutExclusiveLock (3 ms)
> ...
> [ RUN      ] TestMockImageRefreshRequest.OpenObjectMapError
> test/librbd/image/test_mock_RefreshRequest.cc:754: Failure
> Value of: update_features(ictx, (1<<6), false)
>   Actual: -22
> Expected: 0
> [  FAILED  ] TestMockImageRefreshRequest.OpenObjectMapError (3 ms)
>

**Comment by s09816** [comment]
@dillaman  please help to check

**Comment by s09816** [comment]
@dillaman please help to check

**Comment by dillaman** [comment]
@s09816 Fails new test case ```TestMockImageRefreshRequest.DisableExclusiveLockWhileAcquiringLock```

**Comment by s09816** [comment]
@dillaman I have updated

**Pull request merged**